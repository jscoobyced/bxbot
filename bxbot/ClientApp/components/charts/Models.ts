export interface Pairing {
    timestamp: number;
    low: number;
    high: number;
    close: number;
    open: number;
}

export interface FetchDataState {
    pairings: Pairing[];
    loading: boolean;
    url: string;
}
