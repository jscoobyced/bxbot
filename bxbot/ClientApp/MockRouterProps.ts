import { RouteComponentProps } from 'react-router';
import { match } from 'react-router-dom';
import { UnregisterCallback, Href, LocationListener } from 'history';

export class MockRouterProps {
    public static listener: LocationListener = () => { };

    public static getMockRouterProps<P>(data: P) {

        let location = {
            hash: "",
            key: "",
            pathname: "",
            search: "",
            state: {}
        };

        let props: RouteComponentProps<P> = {
            match: {
                isExact: true,
                params: data,
                path: "",
                url: ""
            },
            location: location,
            history: {
                length: 2,
                action: "POP",
                location: location,
                push: () => { },
                replace: () => { },
                go: (num) => { },
                goBack: () => { },
                goForward: () => { },
                block: (t) => {
                    let temp: UnregisterCallback = () => { };
                    return temp;
                },
                createHref: (t) => {
                    var temp: Href = "";
                    return temp;
                },
                listen: (t) => {
                    let temp: UnregisterCallback = () => { };
                    return temp;
                }

            },
            staticContext: {
            }
        };

        return props;
    }
}