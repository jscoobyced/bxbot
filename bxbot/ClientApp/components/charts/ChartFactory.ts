import { } from 'google.visualization';

export class ChartFactory {
    public arrayToDataTable(data: Array<number>): google.visualization.DataTable {
        return google.visualization.arrayToDataTable(data, true);
    }

    public drawCandleChart(data: google.visualization.DataTable, chartElement: string, options: any) {
        const chart = new google.visualization.CandlestickChart(document.getElementById(chartElement) as Element);
        chart.draw(data, options);
    }
}