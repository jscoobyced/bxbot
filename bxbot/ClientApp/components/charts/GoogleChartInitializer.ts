import { GoogleCharts } from 'google-charts';

export class GoogleChartInitializer {
    private static Ready: boolean = false;

    public constructor() {
        GoogleCharts.load(this.drawChart);
    }

    public static IsReady(): boolean {
        return GoogleChartInitializer.Ready;
    }

    private drawChart() {
        GoogleChartInitializer.Ready = true;
    }
}