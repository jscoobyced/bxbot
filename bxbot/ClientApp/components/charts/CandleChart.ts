import { ChartFactory } from './ChartFactory';

export class CandleChart {
    private _chartData: Array<any>;
    private _chartElement: string;
    private _chartFactory: ChartFactory;

    public constructor(chartData: Array<any>, chartElement: string) {
        this._chartData = chartData;
        this._chartElement = chartElement;
        this._chartFactory = new ChartFactory();
    }

    public drawChart() {
        const data = this._chartFactory.arrayToDataTable(this._chartData);
        const options = {
            legend: 'none',
            height: 800,
            candlestick: {
                fallingColor: { strokeWidth: 0, fill: '#a52714' },
                risingColor: { strokeWidth: 0, fill: '#0f9d58' },
                hollowIsRising: true
            },
            colors: ['black'],
            seriesType: 'candlesticks',
            series: {
                1: {
                    type: 'area',
                    lineWidth: 1,
                    color: '#FFCDD2'
                },
                2: {
                    type: 'area',
                    lineWidth: 1,
                    color: '#90CAF9'
                },
                3: {
                    type: 'line',
                    lineWidth: 1
                }
            }
        };

        this._chartFactory.drawCandleChart(data, this._chartElement, options);
    }
}