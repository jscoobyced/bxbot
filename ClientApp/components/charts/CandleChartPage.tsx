import * as React from 'react';
import { RouteComponentProps } from 'react-router';
import { Pairing, FetchDataState } from './Models';
import { CandleChart } from './CandleChart';

export class CandleChartPage extends React.Component<RouteComponentProps<{}>, FetchDataState> {

    private _candleChart: CandleChart = new CandleChart();
    private _chartElement: string = "candle_div"

    constructor() {
        super();
        this.state = { pairings: [], loading: true };
    }

    public componentDidMount() {
        fetch('api/Data/pairing/1/5')
            .then(response => response.json() as Promise<Pairing[]>)
            .then(data => {
                this._candleChart.init(data, this._chartElement);
                this.setState({ loading: false });
            });
    }

    public render() {
        let contents: JSX.Element = <span></span>;
        if (this.state.loading) {
            contents = <p><em>Loading...</em></p>
        }
        else {
            this._candleChart.renderChart();
        }

        return <div>
            <h1>BTC on bx.in.th</h1>
            <div id={this._chartElement}></div>
            {contents}
        </div>;
    }
}