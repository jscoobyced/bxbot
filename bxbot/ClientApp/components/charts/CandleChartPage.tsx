import * as React from 'react';
import { RouteComponentProps } from 'react-router';
import { DateUtil } from '../../utils/DateUtils';
import { Pairing, FetchDataState } from './Models';
import { CandleChart } from './CandleChart';
import { GoogleChartInitializer } from './GoogleChartInitializer';
import { BollingerBand } from "../../utils/Bollinger";
import { DataFormatter } from './DataFormatter';

export class CandleChartPage extends React.Component<RouteComponentProps<{}>, FetchDataState> {
    private static ChartElement: string = "candle_div";
    private _candleChart: CandleChart = new CandleChart([], "");
    private _bollingerSize: number = 20;
    private _dataFormatter: DataFormatter = new DataFormatter();
    private _chartData: any;

    constructor() {
        super();
        this.state = { pairings: [], loading: true, url: 'api/Data/pairing/1/5' };
    }

    public componentDidMount() {
        fetch(this.state.url)
            .then(response => response.json() as Promise<Pairing[]>)
            .then(data => {
                this._chartData = this._dataFormatter.formatBollingerData(data, this._bollingerSize);
                if (!GoogleChartInitializer.IsReady()) {
                    new GoogleChartInitializer().Init(this.drawChart);
                } else {
                    this.drawChart();
                }
            });
    }

    private drawChart = () => {
        this._candleChart = new CandleChart(this._chartData, CandleChartPage.ChartElement);
        this.setState({ loading: false });
    }

    public render(): JSX.Element {
        let contents: JSX.Element = <span></span>;
        if (this.state.loading) {
            contents = <p><em>Loading...</em></p>
        }
        else {
            this._candleChart.drawChart();
        }

        return <div>
            <h1>BTC on bx.in.th</h1>
            <div id={CandleChartPage.ChartElement}></div>
            {contents}
        </div>;
    }
}