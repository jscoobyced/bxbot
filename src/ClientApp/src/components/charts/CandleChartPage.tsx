import * as React from 'react';
import { CandleChart } from './CandleChart';
import { ICandleChartDataService } from './CandleChartDataService';
import { CandleChartDrawer } from './CandleChartDrawer';
import { DataFormatter } from './DataFormatter';
import { GoogleChartInitializer } from './GoogleChartInitializer';
import { CandleChartPageState, CandleChartPageProps } from './Models';

export class CandleChartPage extends React.Component<CandleChartPageProps, CandleChartPageState> {
    private static readonly ChartElement = "candle_div";
    private candleChart: CandleChartDrawer = new CandleChartDrawer([], "");
    private readonly bollingerSize = 20;
    private readonly dataFormatter: DataFormatter = new DataFormatter();
    private chartData: any;
    private readonly service: ICandleChartDataService;

    constructor(props: CandleChartPageProps, state: CandleChartPageState) {
        super(props, state);
        this.service = props.candleChartDataService;
        this.state = { pairings: [], loading: true };
    }

    public componentDidMount() {
        this.service.fetchCurrencyData(5)
            .then(data => {
                this.chartData = this.dataFormatter.formatBollingerData(data, this.bollingerSize);
                if (!GoogleChartInitializer.IsReady()) {
                    new GoogleChartInitializer().Init(this.drawChart);
                } else {
                    this.drawChart();
                }
            });
    }

    public render(): JSX.Element {
        if (!this.state.loading) {
            this.candleChart.drawChart();
        }
        return <CandleChart
            element={CandleChartPage.ChartElement}
            currency='BTC'
        />;
    }

    private readonly drawChart = () => {
        this.candleChart = new CandleChartDrawer(this.chartData, CandleChartPage.ChartElement);
        this.setState({ loading: false });
    }
}
