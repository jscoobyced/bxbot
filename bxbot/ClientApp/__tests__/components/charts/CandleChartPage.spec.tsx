import { CandleChartPage } from '../../../components/charts/CandleChartPage';
import { MockRouterProps } from '../../../MockRouterProps';
import { } from 'jest';
import * as renderer from 'react-test-renderer';
import * as React from 'react';
import "isomorphic-fetch";

describe('CandleChartPage', () => {
    var routerProps = MockRouterProps.getMockRouterProps<{}>({});
    routerProps.history.block();
    routerProps.history.createHref({});
    routerProps.history.listen(MockRouterProps.listener);

    it('should match snapshot', () => {
        const home = renderer
            .create(<CandleChartPage match={routerProps.match} history={routerProps.history} location={routerProps.location} />)
            .toJSON();
        expect(home).toMatchSnapshot();
    });

    (window as any).JSGoogleChart = (window as any).JSGoogleChart || true;
    it('should draw chart', () => {
        const home = renderer
            .create(<CandleChartPage match={routerProps.match} history={routerProps.history} location={routerProps.location} />)
            .toJSON();
        expect(home).toMatchSnapshot();
    });
});