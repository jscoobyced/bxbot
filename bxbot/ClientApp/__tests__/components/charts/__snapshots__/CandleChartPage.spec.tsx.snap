// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CandleChartPage should draw chart 1`] = `
<div>
  <h1>
    BTC on bx.in.th
  </h1>
  <div
    id="candle_div"
  />
  <p>
    <em>
      Loading...
    </em>
  </p>
</div>
`;

exports[`CandleChartPage should match snapshot 1`] = `
<div>
  <h1>
    BTC on bx.in.th
  </h1>
  <div
    id="candle_div"
  />
  <p>
    <em>
      Loading...
    </em>
  </p>
</div>
`;
